dartagnan/src/main/java/com/dat3m/dartagnan/analysis/graphRefinement/graphs/eventGraph/stat/CartesianGraph.java

```diff
@@ -0,0 +1,81 @@
+package com.dat3m.dartagnan.analysis.graphRefinement.graphs.eventGraph.stat;
+
+import com.dat3m.dartagnan.analysis.graphRefinement.util.EdgeDirection;
+import com.dat3m.dartagnan.verification.model.Edge;
+import com.dat3m.dartagnan.verification.model.EventData;
+import com.dat3m.dartagnan.verification.model.ExecutionModel;
+import com.dat3m.dartagnan.wmm.filter.FilterAbstract;
+import com.dat3m.dartagnan.wmm.relation.base.stat.RelCartesian;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Stream;
+
+public class CartesianGraph extends StaticEventGraph {
+
+    private final List<EventData> firstEvents;
+    private final List<EventData> secondEvents;
+
+    private final FilterAbstract first;
+    private final FilterAbstract second;
+
+    public CartesianGraph(RelCartesian rel) {
+        this(rel.getFirstFilter(), rel.getSecondFilter());
+    }
+
+    public CartesianGraph(FilterAbstract first, FilterAbstract second) {
+        this.first = first;
+        this.second = second;
+
+        firstEvents = new ArrayList<>();
+        secondEvents = new ArrayList<>();
+    }
+
+    @Override
+    public boolean contains(EventData a, EventData b) {
+        return first.filter(a.getEvent()) && second.filter(b.getEvent());
+    }
+
+    @Override
+    public void constructFromModel(ExecutionModel model) {
+        super.constructFromModel(model);
+        firstEvents.clear();
+        secondEvents.clear();
+
+        for (EventData e : model.getEventList()) {
+            if (first.filter(e.getEvent()))
+                firstEvents.add(e);
```

Comment:
add `{body}`

Even if the body has a single line, avoid not using `{}` because someone can add more lines without realising the missing brackets and that might cause problems (I have seen this causing problems in other projects)